torch/csrc/distributed/c10d/HealthcheckNCCL.hpp

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#ifdef USE_C10D_NCCL
+
+#include <c10/cuda/CUDAStream.h>
+#include <c10/macros/Export.h>
+#include <torch/csrc/distributed/c10d/Backend.hpp>
+#include <torch/csrc/distributed/c10d/Healthcheck.hpp>
+#include <torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp>
+#include <torch/csrc/distributed/c10d/Store.hpp>
+
+namespace c10d {
+
+class TORCH_API HealthcheckNCCL : public Healthcheck {
+ public:
+  HealthcheckNCCL(
+      const c10::intrusive_ptr<Store>& store,
+      int rank,
+      int worldSize,
+      int localWorldSize,
```

Comment:
a little unclear how you use localWorldSize.

At first i assumed you would control the degree of bisection this way.  E.g. world_size=16, local=4 would give you 4 bisection groups instead of 2.  But the number of 'sides' is hardcoded to 2.  Does that imply local_world_size should always be set to world_size/2, or not?

Reply:
The idea here is to set it to the same number of GPUs on the local host. We want to be able to isolate failures to the specific host so each host is in a group with 1 other host (2 groups per host).

> world_size=16, local=4 would give you 4 bisection groups instead of 2

This actually gives you 8 bisection groups, 4 for each side. Each host however only participates in 2 groups offset from each other.

I.e.

host0:group0,group1
host1:group1,group2
host2:group2:group3
host3:group3,group0

Thus if there's any error that's specific to a single host this can identify that bad host since there should be two failures rather than 1.

Reply:
Added an explicit test for that function that lays it out better